jax/scipy/sparse/linalg.py

```diff
@@ -167,10 +188,377 @@ def cg(A, b, x0=None, *, tol=1e-5, atol=0.0, maxiter=None, M=None):
 
   cg_solve = partial(
       _cg_solve, x0=x0, tol=tol, atol=atol, maxiter=maxiter, M=M)
+
   # real-valued positive-definite linear operators are symmetric
-  real_valued = lambda x: not issubclass(x.dtype.type, np.complexfloating)
+  def real_valued(x):
+    return not issubclass(x.dtype.type, np.complexfloating)
   symmetric = all(map(real_valued, tree_leaves(b)))
   x = lax.custom_linear_solve(
       A, b, solve=cg_solve, transpose_solve=cg_solve, symmetric=symmetric)
   info = None  # TODO(shoyer): return the real iteration count here
   return x, info
+
+
+def _safe_normalize(x, return_norm=False, thresh=None):
+  """
+  Returns the L2-normalized vector (which can be a pytree) x, and optionally
+  the computed norm. If the computed norm is less than the threshold `thresh`,
+  which by default is the machine precision of x's dtype, it will be
+  taken to be 0, and the normalized x to be the zero vector.
+  """
+  norm = _norm_tree(x)
+  dtype = jnp.result_type(*tree_leaves(x))
+  if thresh is None:
+    thresh = jnp.finfo(norm.dtype).eps
+  thresh = thresh.astype(dtype).real
+
+  normalized_x, norm = lax.cond(
+    norm > thresh,
+    lambda y: (_div(y, norm), norm),
+    lambda y: (tree_map(jnp.zeros_like, y), jnp.zeros((), dtype=thresh.dtype)),
+    x,
+  )
+  if return_norm:
+    return normalized_x, norm
+  else:
+    return normalized_x
+
+
+def _project_on_columns(A, v):
+  """
+  Returns A.T.conj() @ v.
+  """
+  v_proj = tree_multimap(
+    lambda X, y: jnp.einsum("...n,...->n", X.conj(), y),
+    A,
+    v,
+  )
+  return tree_reduce(operator.add, v_proj)
+
+
+def _iterative_classical_gram_schmidt(Q, x, iterations=2):
+  """Orthogonalize x against the columns of Q."""
+  # "twice is enough"
+  # http://slepc.upv.es/documentation/reports/str1.pdf
+
+  # This assumes that Q's leaves all have the same dimension in the last
+  # axis.
+  r = jnp.zeros((tree_leaves(Q)[0].shape[-1]))
+  q = x
+
+  for _ in range(iterations):
```

Comment:
You only really need the second iteration if the vectors are nearly linearly dependent. A robust test is to only do the second iteration of ||r||  < 1/sqrt(2) ||x||.

Reply:
Addressed in the latest commit